File: tools/genPdfs.nim
import std / [strformat, strutils, osproc, os, xmltree, htmlparser, streams]
import parsetoml

type
  POKind = enum
    poNone = "other"
    po2006 = "po2006"
    po2014 = "po2014"

  Degree = enum
    dkBachelor = "bsphysik"
    dkMaster = "msphysik"
    dkAstro = "msastro"
    dkLvAndere = "lvandere"

const Prefix = "../public"

proc toMhbPath(degree: string): string = &"mhb_{degree.toLowerAscii}"

proc determineDegree(data: TomlValueRef, course: string): string =
  # same logic as in hugo, i.e. we take the last one we encounter.
  # In Hugo it's more complicated to change this, since we cannot return early,
  # would need to check if result is assigned.
  for deg in data["Degrees"].getElems:
    for locCourse in data[deg.getStr].getElems:
      if course == locCourse.getStr:
        result = deg.getStr
  result = toMhbPath result

proc toOutpath(name: string, poKind: POKind): string =
  createDir(Prefix & "/pdfs" / $poKind)
  result = Prefix / "pdfs" / $poKind / name.parentDir.extractFilename & ".pdf"

proc toCoursePath(poKind, degree, course: string): string =
  # hugo auto replaces spaces in paths by `-`
  let c = course.replace(" ", "-")
  result = Prefix & "/docs" / poKind & degree / "courses" / c / "index.html"

proc toModulePath(poKind, degree, module: string): string =
  # hugo auto replaces spaces in paths by `-`
  let m = module.replace(" ", "-")
  result = Prefix & "/docs" / poKind & degree / m / "index.html"

proc buildPdfs(paths: seq[string]) =
  ## Executes the pandoc commands in parallel using `execProcesses`. The input
  ## files are generated in our case beforehand (but normally will probably
  ## exist). If one or more processes returns an error code > 0, we quit. In
  ## a production environment we should of course handle that in the appropriate
  ## manner instead.
  let res = execProcesses(paths)
  if res != 0:
    quit("Couldn't convert something... (don't quit me..)")

proc genCommands(paths: seq[string], outfile: string): string =
  ## writes out the XML data as "input files" and returns the full pandoc commands
  ## needed to run the conversion. As such we can then call multiple pandoc processes
  ## to do the job for us.
  echo "Building XML file & pandoc command for : ", paths
  var xmlData = ""
  for p in paths:
    let tree = loadHtml(p).findAll("article")
    # there's 2 `<article>` tags in each, they are nested
    doAssert tree.len == 2
    # take the outer tag
    xmlData.add $(tree[0])
  # start a pandoc process and feed the xml data
  let infile = &"/tmp/{outfile.extractFilename}.xml"
  writeFile(infile, $xmlData)
  let cmd = &"pandoc -s {infile} -f html --template mhb_pandoc_template.latex -o {outfile}"
  result = cmd

proc handleCourse(poKind: POKind, course: string): string =
  let courseMap = parseFile(&"../data/{poKind}_course_map.toml")
  let lookupDeg = courseMap.determineDegree course
  # with course and degree build path
  # something like this:
  # for ponone need to ignore poKind.
  result = toCoursePath($poKind & "/", lookupDeg, course)

proc handleModule(data: TomlValueRef, poKind: POKind, degree, module: string) =
  # read module's courses
  var paths = newSeq[string]()
  # add module path
  paths.add toModulePath($poKind & "/", toMhbPath degree, module)
  var commands = newSeq[string]()
  for course in data[module]["CourseList"].getElems:
    # read course map
    let path = handleCourse(poKind, course.getStr)
    # build the commands to generate the outputs
    commands.add genCommands(@[path], outfile = path.toOutpath(poKind))
    # and add path to result
    paths.add path
  echo "Building the commands: ", commands.len
  # now using the commands call `execProcesses` to run them in parallel
  buildPdfs(commands)
  ## TODO: to properly multithread this tool, we'd have to also take care of the
  ## modules themselves, but that's an irrelevant detail for the PoC.

proc handleDegree(poKind: POKind, deg: Degree) =
  # open data file for this degree
  let suffix = if poKind == po2014: $deg & "2" else: $deg
  let data = parseFile(&"../data/mhb_{suffix}.toml")
  for module in data["ModuleList"].getElems:
    data.handleModule(poKind, suffix, module.getStr)

proc handlePO(poKind: POKind) =
  case poKind
  of poNone:
    handleDegree(poNone, dkLvAndere)
  else:
    for deg in Degree:
      if deg == dkLvAndere: continue
      handleDegree(poKind, deg)

for poKind in POKind:
  handlePO(poKind)

File: tools/genPdfs.py
#!/usr/bin/env python3

import toml
import os
from enum import Enum
from subprocess import Popen, PIPE
from pathlib import Path

class POKind(str, Enum):
    poNone = "other"
    po2006 = "po2006"
    po2014 = "po2014"

class Degree(str, Enum):
    dkBachelor = "bsphysik"
    dkMaster = "msphysik"
    dkAstro = "msastro"
    dkLvAndere = "lvandere"

Prefix = "/var/cache/www"

def to_mhb_path(degree: str) -> str:
    return "mhb_{}".format(degree.lower())

def determine_degree(data: dict, course: str) -> str:
    """
    Same logic as in hugo, i.e. we take the last one we encounter.
    In Hugo it's more complicated to change this, since we cannot return early,
    would need to check if result is assigned.
    """
    result = ""
    for deg in data["Degrees"]:
        for locCourse in data[deg]:
            if course == locCourse:
                result = deg
    result = to_mhb_path(result)
    return result

def to_outpath(name: str, poKind: POKind) -> str:
    path = "{}/pdfs/{}".format(Prefix, poKind)
    if not os.path.exists(path):
        os.mkdir(path)
    return os.path.join(path, Path(Path(name).parent.absolute()).name + ".pdf")

def to_course_path(poKind, degree, course: str) -> str:
    # hugo auto replaces spaces in paths by `-`
    c = course.replace(" ", "-")
    return os.path.join(Prefix, "docs", poKind, degree, "courses", c, "index.html")

def to_module_path(poKind, degree, course: str) -> str:
    # hugo auto replaces spaces in paths by `-`
    m = course.replace(" ", "-")
    return os.path.join(Prefix, "docs", poKind, degree, m, "index.html").lower()

def build_pdf(paths: list, outfile: str):

    lcpaths = [p.lower() for p in paths]
    lcoutfile = outfile.lower()

    if not os.path.exists(Prefix + '/pdfs'):
        os.makedirs(Prefix + '/pdfs')

    print("Building PDF for {} storing as {}".format(lcpaths, lcoutfile))
    xmlData = ""
    for p in lcpaths:
        # XML module in python's stdlib cannot handle HTML, thus hack around to extract
        # <article> tag...
        data = open(p).read()
        idx = data.find("<article")
        # there's 2 `<article>` tags in each, they are nested, thus use rfind for closing tag
        idxStop = data.rfind("</article")
        xmlData = xmlData + data[idx : idxStop]
    # start a pandoc process and feed the xml data
    pid = Popen(["pandoc", "-f", "html", "--template",
                 "/var/cache/web-mhb/tools/mhb_pandoc_template.latex", "-o", lcoutfile],
                stdin = PIPE)
    res = pid.communicate(input = xmlData.encode())
    pid.wait()
    if pid.returncode != 0:
        print(pid.stdout)
        print(pid.stderr)
        quit("Failed to convert {}".format(lcpaths))

def handle_course(poKind: POKind, course: str) -> str:
    """
    For each course determine which degree it is part in by the same (somewhat
    silly) way as it is done in Hugo to actually produce the correct results.
    """
    courseMap = toml.load("/var/cache/web-mhb/data/{}_course_map.toml".format(poKind))
    # course may not be in this module's degree. Thus look up degree
    lookupDeg = determine_degree(courseMap, course)
    return to_course_path("{}/".format(poKind), lookupDeg, course)

def handle_module(data: dict, poKind: POKind, degree, module: str):
    """
    For each module check its "CourseList". For each course, generate each individual
    course PDF as well as add the path to that HTML file to a list of paths so we can
    later generate a PDF for the module + its courses.
    """
    paths = []
    # add module path
    paths.append(to_module_path("{}/".format(poKind), to_mhb_path(degree), module))
    for course in data[module]["CourseList"]:
        # read course map
        path = handle_course(poKind, course)
        # build individual course PDF here
        build_pdf([path], outfile = to_outpath(path, poKind))
        # and add path to result
        paths.append(path)
    build_pdf(paths, outfile = to_outpath(paths[0], poKind))

def handleDegree(poKind: POKind, deg: Degree):
    """
    For each degree we read the TOML data file for that degree to lookup the
    modules contained in this degree. This allows us to iterate over all modules
    to generate module PDFs as well as each courses PDF.
    """
    # set possible degree name suffix if PO is 2014
    suffix = "{}2".format(deg) if poKind == POKind.po2014 else "{}".format(deg)
    data = toml.load("/var/cache/web-mhb/data/mhb_{}.toml".format(suffix))
    for module in data["ModuleList"]:
        handle_module(data, poKind, suffix, module)

def handle_PO(poKind: POKind):
    """
    If PO is none, looking at "LV Andere", else make sure to skip that 'Degree'
    in the PO.
    """
    if poKind == POKind.poNone:
        handleDegree(POKind.poNone, Degree.dkLvAndere)
    else:
        for deg in Degree:
            if deg == Degree.dkLvAndere: continue
            handleDegree(poKind, deg)

for poKind in POKind:
  handle_PO(poKind)
